Cmake documentation (#132)
cg-tuwien/Gears-Vk: docs/cmake.md

# CMake

_Gears-Vk_ also supports building with CMake on Linux (*GCC* or *Clang*) and Windows (*MSVC*).

There are currently no pre-built binaries of *Gears-Vk*'s dependencies for Linux included in the repository, so they are built alongside *Gears-Vk* the first time you build it.
However, if there is already an *ASSIMP* build installed on your system (so that it can be found via `find_package(assimp)`), it will be used instead (e.g., on Ubuntu 20.04 `sudo apt-get install libassimp-dev` can be executed to get *ASSIMP* version 5.0.1).
On Windows pre-built binaries of dependencies are included and used when building *Gears-Vk*.

## CLion

### Linux: Ubuntu

> No special settings are required on Linux: CLion can use the bundled cmake, you have to install the following packages on your system to get it work:

Reviewer: ```suggestion
No special settings are required on Linux. CLion can use the bundled cmake. The following packages have to be installed on your system:
```

```
sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
sudo apt update && sudo apt install libassimp-dev g++-11 
sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-11 60 --slave /usr/bin/g++ g++ /usr/bin/g++-11
sudo add-apt-repository ppa:oibaf/graphics-drivers
sudo apt update && sudo apt upgrade -y
sudo apt install libvulkan-dev libvulkan1 mesa-vulkan-drivers vulkan-utils

> # Now verify your vulkan installation:

Reviewer: ```suggestion
# Verify your Vulkan installation:
```
^ Something's missing. Maybe like follows: "Verify the sanity of your Vulkan installation" or something like that?!

vulkaninfo

> # Now install additional vulkan packages:

Reviewer: ```suggestion
# Install additional Vulkan packages:
```

wget -qO - http://packages.lunarg.com/lunarg-signing-key-pub.asc | sudo apt-key add -
sudo wget -qO /etc/apt/sources.list.d/lunarg-vulkan-focal.list http://packages.lunarg.com/vulkan/lunarg-vulkan-focal.list
sudo apt update
sudo apt install vulkan-sdk dpkg-dev libvulkan1-dbgsym vulkan-tools-dbgsym
```

> Now you should be able to build the examples and run it.

Reviewer: ```suggestion
After following these steps, the examples should build and execute successfully.
```

> I only tested this on Ubuntu 20. Please refer to the original sources of vulkan and cmake in order to correctly install the necessary packages for other linux distros.

Reviewer: ```suggestion
These steps have been tested on Ubuntu 20. 
Necessary packages for other Linux distributions can be found in the original sources of Vulkan and CMake.
```

^ Not sure, though, if this makes sense. Because, what are "original sources of Vulkan" and "original sources of CMake"?

### Windows

> Windows needs a little special treatment with CMake. I did not test it with Cygwin or MinGW, so if you try that, be warned that it might not work at all.
> You first have to install the latest Visual Studio and its SDKs.
> Then, assuming you have already installed Vulkan (along with any SDKs and tools) and CMake, and also downloaded the latest CLion, open the project with CLion and go to the Settings menu of CLion:

Reviewer: ```suggestion
Under the assumption that the Visual Studio SDKs, Vulkan SDK, CMake, and CLion are installed, the following settings have to be set in CLion:
```

^ That is not entirely satisfactory, though, because which "Visual Studio SDKs" is this referring to? To the Windows SDK maybe? I.e., either provide complete information, or leave the information out!

The first part of the original text sounded almost like a rant about Windows. The documentation is not the right place for that. Just provide the info about how to get it working here!

Regarding Cygwin and MinGW: It doesn't really help that we know that this has not been tested. I.e., just leave it out and describe an approach that has worked!

You have to adapt the toolchain to Visual Studio and change the default setting to x64, or the code of the framework will not compile.
Then, select the correct toolchain in your CMake settings menu.

![CLion Toolchains Settings](images/clion_toolchains_vs_64bit.png)

> ![CLion CMake Settings](images/clion_cmake_settings.png)

Reviewer: I like these images, but could you mark the settings in a different way please?! :D 
I.e., please draw a rectangle or a (geometric) ellipse around it?!

> You can optionally also set the Build directory to `out`, as this specific name is included in the `.gitignore`, any other default name from CLion is not.

Reviewer: Please avoid addressing the reader with "you"!
